### DISCUSSION TITLE
Allow notifying the user about an upcoming shutdown

### DISCUSSION THREAD

**Discussion by HilkopterBob** [zauberzeug/nicegui#718] [Ideas / Feature Requests]
### Question

hey My app requires making major changes to the codebase on a regular basis. According to my tests, the changes are regularly too big for the nicegui hotreload to cope with. That's why I thought that it would be enough to inform users with a dialog that a shutdown is imminent. So now to my question: is it possible to write a "global" function so that all users on all pages and subpages are informed immediately about the shutdown?

**Comment by falkoschindler** [comment]
Hi @HilkopterBob!
To notify all clients about the app shutting down, I thought we could simply do something like this:

```py
async def bye():
    for client in globals.clients.values():
        with client:
            ui.notify('Goodbye!')

app.on_shutdown(bye)
```

But there's a problem: Async shutdown handlers are not awaited. Right after creating background tasks, the process is stopped. Thus the notification is added to the outbox queue, but is never emitted.

@rodja What do you think? Should we await shutdown handlers? Creating tasks right before stopping the process doesn't make too much sense.

By the way, this might be related to https://github.com/zauberzeug/nicegui/discussions/297.

**Reply by HilkopterBob** [reply]
that would be a great way, @falkoschindler ! 
with an async func we could even make a slight delay and a clock, so the user gets to save the current state before the shutdown actually happens, right?

**Reply by falkoschindler** [reply]
Probably. But we have to see what is possible when the shutdown is already happening.

**Reply by rodja** [reply]
@falkoschindler I kind of remember that we investigated the behaviour and the shutdown-handler from uvicorn is called *after* the connections are dropped. I think it was at the time we implemented #239 (show popup when server connection is lost). But I could be wrong.

**Reply by falkoschindler** [reply]
@rodja You're right. Even if we await the shutdown handlers in `handle_shutdown()` the notification is not transmitted. The connection is disconnected before `handle_shutdown` is done.

**Comment by falkoschindler** [comment]
@HilkopterBob Maybe this is a possible workaround:

1. Use the `uvicorn_reload_*` arguments in `ui.run` to limit the change detection to a single file like ".trigger_reload".
2. Implement a custom change detection using [watchgod](https://pypi.org/project/watchgod/).
3. When a file change is detected, send notifications to the clients.
4. A certain time after notifying the clients, touch the ".trigger_reload" file to actually restart the app.
